internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,249 @@
+	buffer  []lokiEntryWithService
+	entries []lokiEntryWithService
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return NewClientWithOptions(target, &ClientOptions{})
+}
+
+// ClientOptions allows overriding default parameters (e.g. for testing)
+type ClientOptions struct {
+	RequestTimeout    time.Duration
+	MaxRequestEntries int
+}
+
+func NewClientWithOptions(target *plan.LogTarget, options *ClientOptions) *Client {
+	options = fillDefaultOptions(options)
+	c := &Client{
+		options:    options,
+		targetName: target.Name,
+		remoteURL:  target.Location,
+		httpClient: &http.Client{Timeout: options.RequestTimeout},
+		buffer:     make([]lokiEntryWithService, 2*options.MaxRequestEntries),
+	}
+	// c.entries should be backed by the same array as c.buffer
+	c.entries = c.buffer[0:0:len(c.buffer)]
+	return c
+}
+
+func fillDefaultOptions(options *ClientOptions) *ClientOptions {
+	if options.RequestTimeout == 0 {
+		options.RequestTimeout = requestTimeout
+	}
+	if options.MaxRequestEntries == 0 {
+		options.MaxRequestEntries = maxRequestEntries
+	}
+	return options
+}
+
+func (c *Client) Add(entry servicelog.Entry) error {
+	if n := len(c.entries); n >= c.options.MaxRequestEntries {
+		// 'entries' is full - remove the first element to make room
+		// Zero the removed element to allow garbage collection
+		c.entries[0] = lokiEntryWithService{}
+		c.entries = c.entries[1:]
+	}
+
+	if len(c.entries) >= cap(c.entries) {
+		// Copy all the elements to the start of the buffer
+		copy(c.buffer, c.entries)
+
+		// Reset the view into the buffer
+		c.entries = c.buffer[0:len(c.entries):len(c.buffer)]
+
+		// Zero removed elements to allow garbage collection
+		for i := len(c.entries); i < len(c.buffer); i++ {
+			c.buffer[i] = lokiEntryWithService{}
+		}
+	}
+
+	c.entries = append(c.entries, lokiEntryWithService{
+		entry:   encodeEntry(entry),
+		service: entry.Service,
+	})
+	return nil
+}
+
+func encodeEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimSuffix(entry.Message, "\n"),
+	}
+}
+
+func (c *Client) Flush(ctx context.Context) error {
+	if len(c.entries) == 0 {
+		return nil // no-op
+	}
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json; charset=utf-8")
+	httpReq.Header.Set("User-Agent", fmt.Sprintf("pebble/%s", cmd.Version))
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+
+	return c.handleServerResponse(resp)
+}
+
+// resetBuffer drops all buffered logs (in the case of a successful send, or an
+// unrecoverable error).
+func (c *Client) resetBuffer() {
+	c.entries = c.entries[:0]
```

Comment:
We have the same issue of entries not being released here.

I'll go ahead and merge this as we've been here for a bit :), but let's please fix this at the next opportunity.